gateway/ln-gateway/tests/authentication_tests.rs

```diff
@@ -0,0 +1,173 @@
+                client_ref.get_balance(pw, payload.clone())
+            })
+            .await
+            .unwrap();
+
+            // Test gateway authentication on `get_deposit_address` function
+            // * `get_deposit_address` with correct password succeeds
+            // * `get_deposit_address` with incorrect password fails
+            let payload = DepositAddressPayload {
+                federation_id: federation_id.clone(),
+            };
+            test_auth(&gw_password, move |pw| {
+                client_ref.get_deposit_address(pw, payload.clone())
+            })
+            .await
+            .unwrap();
+
+            // Test gateway authentication on `deposit` function
+            // * `deposit` with correct password succeeds
+            // * `deposit` with incorrect password fails
+            let (proof, tx) = bitcoin
+                .send_and_mine_block(
+                    &bitcoin.get_new_address().await,
+                    bitcoin::Amount::from_btc(1.0).unwrap(),
+                )
+                .await;
+            let payload = DepositPayload {
+                federation_id: federation_id.clone(),
+                txout_proof: proof,
+                transaction: tx,
+            };
+            test_auth(&gw_password, move |pw| {
+                client_ref.deposit(pw, payload.clone())
+            })
+            .await
+            .unwrap();
+
+            // Test gateway authentication on `withdraw` function
+            // * `withdraw` with correct password succeeds
+            // * `withdraw` with incorrect password fails
+            let payload = WithdrawPayload {
+                federation_id,
+                amount: bitcoin::Amount::from_sat(100),
+                address: bitcoin.get_new_address().await,
+            };
+            test_auth(&gw_password, |pw| client_ref.withdraw(pw, payload.clone()))
+                .await
+                .unwrap();
+        },
+    )
+    .await?;
+
+    Ok(())
+}
+
+/// Test that a given endpoint/functionality of func fails with the wrong
+/// password but works with the correct one
+async fn test_auth<Fut>(gw_password: &str, func: impl Fn(String) -> Fut) -> anyhow::Result<()>
+where
+    Fut: Future<Output = Result<Response, Error>>,
+{
+    assert_eq!(
+        retry(
+            "fn".to_string(),
+            || async {
+                func(format!("foobar{gw_password}"))
+                    .await
+                    .map_err(|e| anyhow::anyhow!(e))
+            },
+            Duration::from_secs(1),
+            3,
+        )
+        .await?
+        .status(),
+        401
+    );
+    assert_ne!(
+        retry(
+            "fn".to_string(),
+            || async {
+                func(gw_password.to_string())
+                    .await
+                    .map_err(|e| anyhow::anyhow!(e))
+            },
+            Duration::from_secs(1),
+            3,
+        )
+        .await?
+        .status(),
+        401
```

Comment:
`assert_eq(200)` tests for success on authentication, AND success in whatever authenticated api we are currently checking. Since we cannot be sure that our API succeeds after successful auth, we did not assert a `200`.

Reply:
Same discussion, months apart :) https://github.com/fedimint/fedimint/pull/1023#discussion_r1036496046